src/extension/type_scheme.rs

```diff
@@ -0,0 +1,269 @@
+//! Polymorphic type schemes for [OpDef]s.
+//! The type scheme declares a number of TypeParams; any TypeArgs fitting those,
+//! produce a FunctionType for the Op by substitution.
+//!
+//! [OpDef]: super::OpDef
+
+use crate::types::type_param::{check_type_args, TypeArg, TypeParam};
+use crate::types::FunctionType;
+
+use super::{ExtensionRegistry, SignatureError};
+
+/// A polymorphic type scheme for an [OpDef]
+///
+/// [OpDef]: super::OpDef
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
+pub struct OpDefTypeScheme {
+    /// The declared type parameters, i.e., every instantiation ([ExternalOp]) must provide [TypeArg]s for these
+    ///
+    /// [ExternalOp]: crate::ops::custom::ExternalOp
+    pub params: Vec<TypeParam>,
+    /// Template for the Op type. May contain variables up to length of [OpDefTypeScheme::params]
+    body: FunctionType,
+}
+
+impl OpDefTypeScheme {
+    /// Create a new OpDefTypeScheme.
+    /// The [ExtensionRegistry] should be the same (or a subset) of that which will later
+    /// be used to create operations; at this point we only need the types.
+    ///
+    /// #Errors
+    /// Validates that all types in the schema are well-formed and all variables in the body
+    /// are declared with [TypeParam]s that guarantee they will fit.
+    pub fn new(
+        params: impl Into<Vec<TypeParam>>,
+        body: FunctionType,
+        extension_registry: &ExtensionRegistry,
+    ) -> Result<Self, SignatureError> {
+        let params = params.into();
+        body.validate(extension_registry, &params)?;
+        Ok(Self { params, body })
+    }
+
+    pub(super) fn compute_signature(
+        &self,
+        args: &[TypeArg],
+        extension_registry: &ExtensionRegistry,
+    ) -> Result<FunctionType, SignatureError> {
+        check_type_args(args, &self.params)?;
+        Ok(self.body.substitute(extension_registry, args))
```

Comment:
We need to validate that the `args` themselves are closed before we call `substitute`

Reply:
Yeah ok, done - with a comment, 'coz this'll go when/if we make Hugr's polymorphic, but the check in `hugr/validate.rs` should be able to stay (using the Hugr's bound variables when it has them)